orchagent/saihelper.cpp

```diff
@@ -803,3 +831,270 @@ void handleSaiFailure(bool abort_on_failure)
+    initSaiRedisCounterEmptyParameter(flex_counter_group_param.plugin_name);
+    initSaiRedisCounterEmptyParameter(flex_counter_group_param.plugins);
+}
+
+static inline void initSaiRedisCounterParameterFromString(sai_s8_list_t &sai_s8_list, const std::string &str)
+{
+    if (str.length() > 0)
+    {
+        sai_s8_list.list = (int8_t*)const_cast<char *>(str.c_str());
+        sai_s8_list.count = (uint32_t)str.length();
+    }
+    else
+    {
+        sai_s8_list.list = nullptr;
+        sai_s8_list.count = 0;
+    }
+}
+
+static inline void notifySyncdCounterOperation(bool is_gearbox, const sai_attribute_t &attr)
+{
+    if (sai_switch_api == nullptr)
+    {
+        // This can happen during destruction of the orchagent daemon.
+        SWSS_LOG_ERROR("sai_switch_api is NULL");
+        return;
+    }
+
+    if (!is_gearbox)
+    {
+        sai_switch_api->set_switch_attribute(gSwitchId, &attr);
+    }
+    else
+    {
+        for (auto gearbox_oid : gGearboxOids)
+        {
+            sai_switch_api->set_switch_attribute(gearbox_oid, &attr);
+        }
+    }
+}
+
+static inline void operateFlexCounterDbSingleField(std::vector<FieldValueTuple> &fvTuples,
+                                            const string &field, const string &value)
+{
+    if (!field.empty() && !value.empty())
+    {
+        fvTuples.emplace_back(field, value);
+    }
+}
+
+static inline void operateFlexCounterGroupDatabase(const string &group,
+                                            const string &poll_interval,
+                                            const string &stats_mode,
+                                            const string &plugin_name,
+                                            const string &plugins,
+                                            const string &operation,
+                                            bool is_gearbox)
+{
+    std::vector<FieldValueTuple> fvTuples;
+    auto &flexCounterGroupTable = is_gearbox ? gGearBoxFlexCounterGroupTable : gFlexCounterGroupTable;
+
+    operateFlexCounterDbSingleField(fvTuples, POLL_INTERVAL_FIELD, poll_interval);
+    operateFlexCounterDbSingleField(fvTuples, STATS_MODE_FIELD, stats_mode);
+    operateFlexCounterDbSingleField(fvTuples, plugin_name, plugins);
+    operateFlexCounterDbSingleField(fvTuples, FLEX_COUNTER_STATUS_FIELD, operation);
+
+    flexCounterGroupTable->set(group, fvTuples);
+}
+void setFlexCounterGroupParameter(const string &group,
+                                  const string &poll_interval,
+                                  const string &stats_mode,
+                                  const string &plugin_name,
+                                  const string &plugins,
+                                  const string &operation,
+                                  bool is_gearbox)
+{
+    if (gTraditionalFlexCounter)
+    {
+        operateFlexCounterGroupDatabase(group, poll_interval, stats_mode, plugin_name, plugins, operation, is_gearbox);
+        return;
+    }
```

Comment:
so, normally in the OOP, such cases is resolved in proper way, by having base abstract class, and pointer in that class in OA, then base on command line in OA whether we want to use traditional counters or new one, just one of 2 objects is createed and populate the pointer, then all operations are the same, add field to counder, add counter remove counter etc, and in all entire code there is no need for gTraditionalFlexCounter) check, and i saw is checked in multiple places, and old and new counter realise the same functionality at the end, but im aware that at this stage this conversion without refactoring OA maybe not possible, it should be designed from the start

Reply:
yea, look all the below functions, they all contain the same check, and adding new code could confuse unaware programmer, so moving this to 2 classes with one new and second ol would be great solution to that